File: vouch/robotics/embodiment.py
"""
Cross-embodiment identity continuity: one accountable agent across robot bodies.

An AI agent (a "mind": a policy with its own Vouch identity) can run on one robot
body today and a different body tomorrow. This makes that continuous and
accountable. An embodiment credential binds the agent identity to a specific body
(a hardware-rooted robot identity) and that body's hardware root for a period,
signed by the agent's own persistent key. Linking each embodiment to the previous
forms a continuity chain a verifier walks to confirm the same accountable agent
persisted across bodies, re-binding to each body's hardware root as it moved. A
fork check confirms the agent was never actively embodied in two bodies at once.

This is the inverse of the ownership custody chain: there one body passes between
owners; here one mind passes between bodies, and the constant that signs every link
is the agent identity itself.

This is the open layer: plain signed embodiment credentials, continuity-chain
verification, and software fork detection. Managed key custody and fleet-scale
migration are out of scope for the open layer.
"""

from __future__ import annotations

from datetime import datetime, timedelta, timezone
from typing import Any, Dict, List, Optional

from ._signing import attach_proof
from .identity import RoboticsError

VC_CONTEXT_V2 = "https://www.w3.org/ns/credentials/v2"
VOUCH_CONTEXT_V1 = "https://vouch-protocol.com/contexts/v1"
EMBODIMENT_TYPE = "AgentEmbodimentCredential"


# ---------------------------------------------------------------------------
# Embodiment credential + continuity chain
# ---------------------------------------------------------------------------


def build_embodiment(
    agent_signer: Any,
    *,
    agent_did: str,
    body_did: str,
    body_hardware_root: str,
    from_body: Optional[str] = None,
    embodied_at: Optional[datetime] = None,
    valid_seconds: Optional[int] = None,
) -> Dict[str, Any]:
    """
    Build a signed embodiment credential: the agent `agent_did` authorizes running
    on `body_did`, re-binding to that body's hardware root `body_hardware_root`.
    Signed by the agent's own persistent key, so the whole continuity chain is
    signed by one accountable identity. `from_body` links this embodiment to the
    body the agent left, forming the chain. `valid_seconds`, when given, bounds the
    active window (used by fork detection).
    """
    if not agent_did or not body_did or not body_hardware_root:
        raise RoboticsError("agent_did, body_did, and body_hardware_root are required")
    issued = (embodied_at or datetime.now(timezone.utc)).astimezone(timezone.utc)
    subject: Dict[str, Any] = {
        "id": agent_did,
        "body": body_did,
        "bodyHardwareRoot": body_hardware_root,
    }
    if from_body is not None:
        subject["fromBody"] = from_body

    credential: Dict[str, Any] = {
        "@context": [VC_CONTEXT_V2, VOUCH_CONTEXT_V1],
        "type": ["VerifiableCredential", EMBODIMENT_TYPE],
        "issuer": agent_did,
        "validFrom": _iso(issued),
        "credentialSubject": subject,
    }
    if valid_seconds is not None:
        credential["validUntil"] = _iso(issued + timedelta(seconds=valid_seconds))
    return attach_proof(credential, agent_signer)


def verify_embodiment(
    credential: Dict[str, Any],
    agent_public_key: Any,
) -> "tuple[bool, Optional[Dict[str, Any]]]":
    """
    Verify an embodiment credential: the agent's proof and that the issuer is the
    agent itself (a mind authorizes its own embodiment). Returns (ok, subject).
    """
    ok, subject = _verify_typed(credential, agent_public_key, EMBODIMENT_TYPE)
    if not ok:
        return False, None
    if not subject.get("body") or not subject.get("bodyHardwareRoot"):
        return False, None
    if credential.get("issuer") != subject.get("id"):
        return False, None
    return True, subject


def verify_continuity_chain(
    embodiments: List[Dict[str, Any]],
    agent_public_key: Any,
    *,
    origin_body: Optional[str] = None,
) -> "tuple[bool, Optional[str]]":
    """
    Verify an ordered list of embodiment credentials forms a valid continuity chain
    for one agent: every link verifies under the SAME agent key (the persistent
    mind), each link's `fromBody` matches the previous link's `body`, and (when
    given) the first `fromBody` is `origin_body`. Returns (ok, current_body).
    """
    expected_from = origin_body
    current_body: Optional[str] = origin_body
    for embodiment in embodiments:
        ok, subject = verify_embodiment(embodiment, agent_public_key)
        if not ok:
            return False, None
        if expected_from is not None and subject.get("fromBody") != expected_from:
            return False, None
        current_body = subject.get("body")
        expected_from = current_body
    return True, current_body


# ---------------------------------------------------------------------------
# Fork detection (a mind cannot be actively embodied in two bodies at once)
# ---------------------------------------------------------------------------


def check_no_fork(
    embodiments: List[Dict[str, Any]],
) -> "tuple[bool, Optional[Dict[str, str]]]":
    """
    Confirm no two embodiments place the agent in different bodies with overlapping
    active windows. Each embodiment is active from `validFrom` to `validUntil` (a
    missing `validUntil` is treated as open-ended). Two embodiments on different
    bodies whose windows overlap are a fork. Returns (ok, conflict) where conflict,
    when present, names the two conflicting bodies.
    """
    windows = []
    for embodiment in embodiments:
        subject = embodiment.get("credentialSubject") or {}
        body = subject.get("body")
        start = _parse_iso(embodiment.get("validFrom"))
        if body is None or start is None:
            return False, None
        end = _parse_iso(embodiment.get("validUntil"))  # None -> open-ended
        windows.append((body, start, end))

    for i in range(len(windows)):
        body_i, start_i, end_i = windows[i]
        for j in range(i + 1, len(windows)):
            body_j, start_j, end_j = windows[j]
            if body_i == body_j:
                continue
            if _overlaps(start_i, end_i, start_j, end_j):
                return False, {"bodyA": body_i, "bodyB": body_j}
    return True, None


def _overlaps(
    start_a: datetime,
    end_a: Optional[datetime],
    start_b: datetime,
    end_b: Optional[datetime],
) -> bool:
    # Half-open intervals [start, end); a missing end is +infinity. A clean handover
    # sets one window's end to the next window's start, which does not overlap.
    a_before_b = end_a is not None and end_a <= start_b
    b_before_a = end_b is not None and end_b <= start_a
    return not (a_before_b or b_before_a)


# ---------------------------------------------------------------------------
# Shared helpers
# ---------------------------------------------------------------------------


def _verify_typed(
    credential: Dict[str, Any],
    public_key: Any,
    expected_type: str,
) -> "tuple[bool, Dict[str, Any]]":
    from vouch import data_integrity
    from vouch.verifier import _coerce_ed25519_public_key

    type_field = credential.get("type") or []
    if isinstance(type_field, str):
        type_field = [type_field]
    if expected_type not in type_field:
        return False, {}
    resolved = _coerce_ed25519_public_key(public_key) if public_key is not None else None
    if resolved is None:
        return False, {}
    try:
        if not data_integrity.verify_proof(credential, resolved):
            return False, {}
    except ValueError:
        return False, {}
    return True, credential.get("credentialSubject") or {}


def _iso(dt: datetime) -> str:
    return dt.astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


def _parse_iso(value: Optional[str]) -> Optional[datetime]:
    if not value:
        return None
    try:
        return datetime.strptime(value, "%Y-%m-%dT%H:%M:%SZ").replace(tzinfo=timezone.utc)
    except (ValueError, TypeError):
        return None


__all__ = [
    "EMBODIMENT_TYPE",
    "build_embodiment",
    "verify_embodiment",
    "verify_continuity_chain",
    "check_no_fork",
]

File: packages/sdk-ts/src/robotics/embodiment.ts
/**
 * Cross-embodiment identity continuity (TypeScript): one accountable agent
 * across robot bodies.
 *
 * Mirrors `vouch/robotics/embodiment.py` with byte-identical output. An AI agent
 * (a "mind": a policy with its own Vouch identity) can run on one robot body
 * today and a different body tomorrow. This makes that continuous and
 * accountable. An embodiment credential binds the agent identity to a specific
 * body (a hardware-rooted robot identity) and that body's hardware root for a
 * period, signed by the agent's own persistent key. Linking each embodiment to
 * the previous forms a continuity chain a verifier walks to confirm the same
 * accountable agent persisted across bodies, re-binding to each body's hardware
 * root as it moved. A fork check confirms the agent was never actively embodied
 * in two bodies at once.
 *
 * This is the inverse of the ownership custody chain: there one body passes
 * between owners; here one mind passes between bodies, and the constant that
 * signs every link is the agent identity itself.
 *
 * This is the open layer: plain signed embodiment credentials, continuity-chain
 * verification, and software fork detection. Managed key custody and fleet-scale
 * migration are out of scope for the open layer.
 */

import * as crypto from 'crypto';

import { verifyProof } from '../data-integrity';
import type { Signer } from '../signer';
import { VC_CONTEXT_V2, VOUCH_CONTEXT_V1 } from '../vc';

import { RoboticsError } from './identity';

export const EMBODIMENT_TYPE = 'AgentEmbodimentCredential';

function iso(d: Date): string {
  return d.toISOString().replace(/\.\d{3}Z$/, 'Z');
}

function parseIso(value: unknown): Date | undefined {
  if (typeof value !== 'string' || !value) return undefined;
  const m = /^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})Z$/.exec(value);
  if (!m) return undefined;
  const ms = Date.UTC(
    Number(m[1]),
    Number(m[2]) - 1,
    Number(m[3]),
    Number(m[4]),
    Number(m[5]),
    Number(m[6])
  );
  if (Number.isNaN(ms)) return undefined;
  return new Date(ms);
}

/**
 * Verify a typed embodiment credential: the expected type is present and the
 * proof verifies under `publicKey`. Returns { ok, subject }.
 */
function verifyTyped(
  credential: Record<string, unknown>,
  publicKey: crypto.KeyObject,
  expectedType: string
): { ok: boolean; subject: Record<string, unknown> } {
  const typeField = credential.type;
  const types = Array.isArray(typeField) ? typeField : [typeField];
  if (!types.includes(expectedType)) return { ok: false, subject: {} };

  if (publicKey === null || publicKey === undefined) return { ok: false, subject: {} };
  try {
    if (!verifyProof(credential, publicKey)) return { ok: false, subject: {} };
  } catch {
    return { ok: false, subject: {} };
  }
  const subject = (credential.credentialSubject ?? {}) as Record<string, unknown>;
  return { ok: true, subject };
}

// ---------------------------------------------------------------------------
// Embodiment credential + continuity chain
// ---------------------------------------------------------------------------

export interface BuildEmbodimentOptions {
  agentDid: string;
  bodyDid: string;
  bodyHardwareRoot: string;
  fromBody?: string;
  embodiedAt?: Date;
  validSeconds?: number;
}

/**
 * Build a signed embodiment credential: the agent `agentDid` authorizes running
 * on `bodyDid`, re-binding to that body's hardware root `bodyHardwareRoot`.
 * Signed by the agent's own persistent key, so the whole continuity chain is
 * signed by one accountable identity. `fromBody` links this embodiment to the
 * body the agent left, forming the chain. `validSeconds`, when given, bounds the
 * active window (used by fork detection).
 */
export async function buildEmbodiment(
  agentSigner: Signer,
  opts: BuildEmbodimentOptions
): Promise<Record<string, unknown>> {
  if (!opts.agentDid || !opts.bodyDid || !opts.bodyHardwareRoot) {
    throw new RoboticsError('agentDid, bodyDid, and bodyHardwareRoot are required');
  }
  const issued = opts.embodiedAt ?? new Date();
  const subject: Record<string, unknown> = {
    id: opts.agentDid,
    body: opts.bodyDid,
    bodyHardwareRoot: opts.bodyHardwareRoot,
  };
  if (opts.fromBody !== undefined) {
    subject.fromBody = opts.fromBody;
  }

  const credential: Record<string, unknown> = {
    '@context': [VC_CONTEXT_V2, VOUCH_CONTEXT_V1],
    type: ['VerifiableCredential', EMBODIMENT_TYPE],
    issuer: opts.agentDid,
    validFrom: iso(issued),
    credentialSubject: subject,
  };
  if (opts.validSeconds !== undefined) {
    credential.validUntil = iso(new Date(issued.getTime() + opts.validSeconds * 1000));
  }
  return agentSigner.attachProof(credential);
}

/**
 * Verify an embodiment credential: the agent's proof and that the issuer is the
 * agent itself (a mind authorizes its own embodiment). Returns { ok, subject }.
 */
export function verifyEmbodiment(
  credential: Record<string, unknown>,
  agentPublicKey: crypto.KeyObject
): { ok: boolean; subject?: Record<string, unknown> } {
  const { ok, subject } = verifyTyped(credential, agentPublicKey, EMBODIMENT_TYPE);
  if (!ok) return { ok: false };
  if (!subject.body || !subject.bodyHardwareRoot) return { ok: false };
  if (credential.issuer !== subject.id) return { ok: false };
  return { ok: true, subject };
}

export interface VerifyContinuityChainOptions {
  originBody?: string;
}

/**
 * Verify an ordered list of embodiment credentials forms a valid continuity
 * chain for one agent: every link verifies under the SAME agent key (the
 * persistent mind), each link's `fromBody` matches the previous link's `body`,
 * and (when given) the first `fromBody` is `originBody`. Returns
 * { ok, currentBody }.
 */
export function verifyContinuityChain(
  embodiments: Array<Record<string, unknown>>,
  agentPublicKey: crypto.KeyObject,
  opts: VerifyContinuityChainOptions = {}
): { ok: boolean; currentBody?: string } {
  let expectedFrom = opts.originBody;
  let currentBody: string | undefined = opts.originBody;
  for (const embodiment of embodiments) {
    const { ok, subject } = verifyEmbodiment(embodiment, agentPublicKey);
    if (!ok || subject === undefined) return { ok: false };
    if (expectedFrom !== undefined && subject.fromBody !== expectedFrom) return { ok: false };
    currentBody = subject.body as string;
    expectedFrom = currentBody;
  }
  return { ok: true, currentBody };
}

// ---------------------------------------------------------------------------
// Fork detection (a mind cannot be actively embodied in two bodies at once)
// ---------------------------------------------------------------------------

export interface CheckNoForkConflict {
  bodyA: string;
  bodyB: string;
}

/**
 * Half-open intervals [start, end); a missing end is +infinity. A clean handover
 * sets one window's end to the next window's start, which does not overlap.
 */
function overlaps(
  startA: Date,
  endA: Date | undefined,
  startB: Date,
  endB: Date | undefined
): boolean {
  const aBeforeB = endA !== undefined && endA.getTime() <= startB.getTime();
  const bBeforeA = endB !== undefined && endB.getTime() <= startA.getTime();
  return !(aBeforeB || bBeforeA);
}

/**
 * Confirm no two embodiments place the agent in different bodies with
 * overlapping active windows. Each embodiment is active from `validFrom` to
 * `validUntil` (a missing `validUntil` is treated as open-ended). Two
 * embodiments on different bodies whose windows overlap are a fork. Returns
 * { ok, conflict } where conflict, when present, names the two conflicting
 * bodies.
 */
export function checkNoFork(
  embodiments: Array<Record<string, unknown>>
): { ok: boolean; conflict?: CheckNoForkConflict } {
  const windows: Array<{ body: string; start: Date; end: Date | undefined }> = [];
  for (const embodiment of embodiments) {
    const subject = (embodiment.credentialSubject ?? {}) as Record<string, unknown>;
    const body = subject.body;
    const start = parseIso(embodiment.validFrom);
    if (typeof body !== 'string' || start === undefined) return { ok: false };
    const end = parseIso(embodiment.validUntil); // undefined -> open-ended
    windows.push({ body, start, end });
  }

  for (let i = 0; i < windows.length; i++) {
    const wi = windows[i];
    for (let j = i + 1; j < windows.length; j++) {
      const wj = windows[j];
      if (wi.body === wj.body) continue;
      if (overlaps(wi.start, wi.end, wj.start, wj.end)) {
        return { ok: false, conflict: { bodyA: wi.body, bodyB: wj.body } };
      }
    }
  }
  return { ok: true };
}
